platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClientsManager.java

```diff
@@ -85,24 +80,42 @@ WorkerDaemonClient reserveIdleClient(DaemonForkOptions forkOptions, List<WorkerD
             Iterator<WorkerDaemonClient> it = clients.iterator();
             while (it.hasNext()) {
                 WorkerDaemonClient candidate = it.next();
-                if (candidate.isCompatibleWith(forkOptions)) {
+                if (candidate.isFailed()) {
+                    emitUnexpectedWorkerFailureWarning(candidate);
                     it.remove();
-                    if (candidate.getLogLevel() != currentLogLevel) {
-                        // TODO: Send a message to workers to change their log level rather than stopping
-                        LOGGER.info("Log level has changed, stopping idle worker daemon with out-of-date log level.");
-                        candidate.stop();
-                    } else {
-                        return candidate;
+                } else {
+                    if (candidate.isCompatibleWith(forkOptions)) {
+                        it.remove();
+                        if (candidate.getLogLevel() != currentLogLevel) {
+                            // TODO: Send a message to workers to change their log level rather than stopping
+                            LOGGER.info("Log level has changed, stopping idle worker daemon with out-of-date log level.");
+                            candidate.stop();
+                        } else {
+                            return candidate;
+                        }
                     }
                 }
             }
             return null;
         }
     }
 
+    private static void emitUnexpectedWorkerFailureWarning(WorkerDaemonClient candidate) {
+        if (candidate.getExitCode().isPresent()) {
+            int exitCode = candidate.getExitCode().get();
+            if (exitCode > 127) {
+                LOGGER.warn("Worker daemon '" + candidate.getDisplayName() + "' exited unexpectedly after being killed with signal " + (exitCode - 128) + ".  This is likely because an external process has killed the worker.");
```

Comment:
I don't think we should assign meaning to any specific range of exit code. It's rare but possible that a program could call `System.exit` with a >127 code, or on Windows I don't think this range means anything, and being forcibly killed by the OS would be a different code than what would be displayed as "signal X".

I think we should just state in all cases "This could be because the OS or another process killed the worker."

Reply:
I understand your point, but I don't agree that we should remove this.  I'm trying to give the user as much information as possible so they can get to the bottom of the behavior.  When these exit codes show up, the reports almost always starts as "workers processes are crashing" which indicates to me that the >127 exit code is not something that most users are aware of.  I think it helps to understand that the processes are being killed rather than exiting for some other "internal" reason (i.e. heap exhaustion, fatal error, `System.exit()`, etc). It _is_ possible that it could be due to something calling `System.exit()` with a value over 127, but I think that would be fairly rare.  I would rather give a little bit of advice that's likely to lead them to a correct understanding of the problem than give them less information on the fear that in exotic cases it would be incorrect.  However, I think the point that Windows doesn't observe this convention is well made.  I've updated the condition so that we only do this when exit code is >127 _and_ we're running on a *nix OS.

Reply:
Sure, that works for me.